File: snippets/python/algo/general/algo_union_find.py
NB_ELEMS = 1000

UF_parent = [-1]*NB_ELEMS
UF_rank = [1]*NB_ELEMS

def Find(a):
	if UF_parent[a] == -1: return a
	UF_parent[a] = Find(UF_parent[a])
	return UF_parent[a]

def Union(a, b):
	a, b = Find(a), Find(b)
	if a==b:return False
	if UF_rank[a] < UF_rank[b]: a, b = b, a
	UF_rank[a], UF_parent[b] = max(UF_rank[a], UF_rank[b]+1), a
	return True

File: snippets/algo/general/algo_union_find.cpp
//@title Union-Find algorithm
//@defines Union Find
//@
const int UF_MAX_NODES = 100*1000;
int uf_parent[UF_MAX_NODES];
int uf_size[UF_MAX_NODES];

struct UF_INIT {
	UF_INIT() {
		fill(uf_parent, uf_parent+UF_MAX_NODES, -1);
		fill(uf_size, uf_size+UF_MAX_NODES, 1);
	}
} test;

int Find(int a) {
	return uf_parent[a] == -1 ? a : (uf_parent[a] = Find(uf_parent[a]));
}
int Union(int a, int b) {
	a = Find(a), b = Find(b);
	if (a != b) {
		if (uf_size[a] < uf_size[b]) {
			swap(a, b);
		}
		uf_parent[b] = a;
		uf_size[a] += uf_size[b];
		return uf_size[a];
	}
	return 0;
}

